### DISCUSSION TITLE
New Sqlalchemy Syntax

### DISCUSSION THREAD

**Discussion by JimScope** [fastapi-users/fastapi-users#999] [Q&A] [answered]
Hello. I have been using fastapi-users with sqlalchemy. But now when migrating to version 10 it uses async. In the example that is in the repo, when I go to query the database, it gives me an AttributeError with session. Any idea what could be wrong?
Here is what i doing [Sample Code](https://pastebin.com/RUC90KBc)

**Comment by frankie567** [comment, marked as answer]
Hello @JimScope!

In your sample code, you are calling the `.scalars` method, which is a special syntax when you want to get single scalar values, like numbers. In SQLAlchemy 1.4/2.x, when you execute a statement, you get a [`Row` objects, which contains the object ORM](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#selecting-orm-entities-and-columns).

It looks like this:

```py
async def get_messages(user: UUID, session: AsyncSession = Depends(get_async_session)):
    statement = select(Message).where(Message.user_id == user).order_by(Message.phone)
    results = await session.execute(statement)
    return [row[0] for row in results.all()]
```

**Reply by JimScope** [reply]
Thanks for highlighting the scalar error for me. But still I face the error that the execute attribute does not exist. Maybe I've done something wrong using get async_session. Excuse my ignorance, but should I use that function as it is in the example for models other than User?

**Reply by frankie567** [reply]
Ok, I think I understand what's going on here: you are calling this function *manually*, *outside* FastAPI, right? If so, this indeed can't work.

Actually, the FastAPI dependency injection system, with all the `Depends` only works inside FastAPI. If you call the dependency function outside, you'll have to manually provide the arguments for it. I give a grasp of this problem here : https://fastapi-users.github.io/fastapi-users/10.0/cookbook/create-user-programmatically/

What I usually do to tackle this is to define *classes*. In the initializer, I can pass the "injection" arguments I can pass either through the FastAPI dependency injection system or manually.

For example:

```py
class MessageRepository:
    def __init___(self, session: AsyncSession = Depends(get_async_session))
        self.session = session

    async def list(self, user: UUID):
        statement = select(Message).where(Message.user_id == user).order_by(Message.phone)
        results = await self.session.execute(statement)
        return [row[0] for row in results.all()]
```

Then, inside a FastAPI router:

```py
app = FastAPI()

@app.get("/messages")
async def list_messages(self, repository: MessageRepository = Depends(MessageRepository), user: User = Depends(current_active_user)):
    messages = await repository.list(user.id)
```

**Reply by JimScope** [reply]
Exactly, thanks @frankie567 . It is also how it is implemented internally in FastAPI-Users

**Reply by frankie567** [reply]
Yes, this is exactly the same pattern with `get_user_db`, `get_user_manager` and so on.
